### PR TITLE
Nock interpreter in Scheme

### PR DESCRIPTION
**Pull request by kohanyirobert**
After writing a Nock interpreter in Groovy #9 I've decided to rewrite it in Scheme. Groovy does not handle tail recursion really well and is bit slow. I've rewritten the Groovy code to use continuation-passing style tail recursion, but it didn't help too much - `StackOverflowError`s all over the place. (I've updated `groovy.md` to reflect the changes.)

The Scheme version benefits from the continuation-passing style recursion. A deep recursion basically runs forever (it does pollutes the stack/heap - don't know which one - but much slower). Anyway, I didn't know too much about CPS before, so it was worth digging into it a bit. The code is a direct rewrite of the Groovy one, no novelty here. Also, I haven't used Scheme since reading the The Little Schemer ages ago, so it might look ugly :)

Sorry for only contributing only useless things, but it was fun.

*goes back figuring out how to implement decrement in nock*

### CODE REVIEW

File: docs/nock/implementations/scheme.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+---
+navhome: /docs
+title: Scheme
+sort: 8
+next: true
+---
+
+# Groovy
```
Comment (galenwp) [inline docs/nock/implementations/scheme.md]: I think you mean `# Scheme`

**Comment by galenwp** [comment]
@philipcmonk is our resident lisp guy — so I'd defer to him on the code.  I think you accidentally copied the Groovy title into the Scheme file.

**Comment by philipcmonk** [comment]
I like it.  Once upon a time I wrote a nock interpreter in Racket.  Similar in a lot of ways, but yours seems cleaner.

https://github.com/philipcmonk/racketnock

**Comment by kohanyirobert** [comment]
@philipcmonk I don't know if it's cleaner or not, but my multiple usage of `(list (list ... )` seems a bit off, however I don't know enough Scheme/LISP in general to know about a better way.

**Comment by philipcmonk** [comment]
Yeah, mine does the same thing, except instead of operating on lists, I use cells (improper lists) at every level.  You need to create a lot of multilevel cell/list structures because nock is a combinator calculus.  I'm not completely sure how you seem to be able to do it with proper lists; does your version give proper lists as its result?  Unless the result is an atom, I mean.

**Comment by kohanyirobert** [comment]
I've started off with Groovy which only have lists. I've wanted to be able to write cells without watching out for proper pair/bracket balance. So I've written a function ([I call it `n`](https://github.com/kohanyirobert/gnock/blob/master/gnock.groovy#L29)) which takes a value and creates a proper noun from it. If the value is an atom it's just returned, if it's a list it creates a new list with proper grouping/pairs ([here's the test case](https://github.com/kohanyirobert/gnock/blob/master/gnock.groovy#L268)).

As the Scheme code is just a rewrite of the Groovy code I've used the same idea. I wanted to be able to write S-expressions where I can omit non-essential brackets/parens. So I'm turning a proper list into an improper one with correct bracket/paren balance ([test case](https://github.com/kohanyirobert/snock/blob/master/snock.ss#L122)).

I'm [using the `n` function to transform the ugly `(list (list ... ` thing](https://github.com/kohanyirobert/snock/blob/master/snock.ss#L100) (as in `(tar (n (list ...`). I remember that I wanted to just use `(tar (cons ...` instead, but it didn't work for some reason ... or maybe I just lost track writing too many `(cons ...` and gave up. Right now it's ugly a bit, but I can still follow it ... kind of (next week from now I'll deny I wrote this :D).

**Comment by philipcmonk** [comment]
Ah, I see it now.  `nock-eval` in mine is equivalent to your `tar` composed with `n`.  It's definitely easy to lose track of `cons`'s - it's always awkward to work against the grain of the language.

I'm not real sure why you ever need `(list` of a single item.  In nock `[a] == a`.  What breaks if you reduce that?

**Comment by kohanyirobert** [comment]
>  In nock [a] == a. What breaks if you reduce that?

Just tried, nothing breaks. Yay!

> I'm not real sure why you ever need (list of a single item.

Artistic freedom I guess. Didn't really think about it. I can just raise an error when the input is `[a]`, but whatever, the code never gets there (unless in the test).

**Comment by philipcmonk** [comment]
Cool, makes sense.

**Pull request merged**